File: Java/051.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Problem051 {

    private static Set<Integer> primesSet;

    public static void sieveOfEratosthenes(int limit) {
        boolean[] isPrime = new boolean[limit + 1];
        Arrays.fill(isPrime, true);
        isPrime[0] = false;
        isPrime[1] = false;

        for (int p = 2; p * p <= limit; p++) {
            if (isPrime[p]) {
                for (int i = p * p; i <= limit; i += p) {
                    isPrime[i] = false;
                }
            }
        }

        primesSet = new HashSet<>();
        for (int i = 2; i <= limit; i++) {
            if (isPrime[i]) {
                primesSet.add(i);
            }
        }
    }

    public static boolean digitReplace(int n) {
        String s = String.valueOf(n);
        Set<Character> uniqueDigits = new HashSet<>();
        for (char c : s.toCharArray()) {
            uniqueDigits.add(c);
        }

        for (char digitToReplace : uniqueDigits) {
            int primeCount = 0;
            for (int i = 0; i <= 9; i++) {
                char replacementDigit = Character.forDigit(i, 10);
                String tempS = s.replace(digitToReplace, replacementDigit);

                // Check for leading zero and length change
                if (tempS.length() != s.length()) {
                    continue;
                }

                int num = Integer.parseInt(tempS);
                if (primesSet.contains(num)) {
                    primeCount++;
                }
            }
            if (primeCount == 8) {
                return true;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        int limit = 10_000_000;
        sieveOfEratosthenes(limit);

        for (int prime : primesSet) {
            // Skip single-digit primes as they cannot form families by replacing digits
            if (prime < 10) {
                continue;
            }
            if (digitReplace(prime)) {
                System.out.println(prime);
                break; // Found the smallest
            }
        }
    }
}

File: Rust/051.rs
use std::collections::HashSet;

fn sieve_of_eratosthenes(limit: usize) -> Vec<u32> {
    let mut is_prime = vec![true; limit + 1];
    if limit >= 0 { is_prime[0] = false; }
    if limit >= 1 { is_prime[1] = false; }

    for p in 2..=limit {
        if p * p > limit { // Optimization: only iterate up to sqrt(limit)
            break;
        }
        if is_prime[p] {
            for multiple in (p * p..=limit).step_by(p) {
                is_prime[multiple] = false;
            }
        }
    }

    let mut primes = Vec::new();
    for (i, &prime) in is_prime.iter().enumerate() {
        if prime {
            primes.push(i as u32);
        }
    }
    primes
}

fn digit_replace(n: u32, primes_set: &HashSet<u32>) -> bool {
    let s = n.to_string();
    let chars: Vec<char> = s.chars().collect();
    let len = chars.len();

    let mut unique_digits_in_n = HashSet::new();
    for &c in &chars {
        unique_digits_in_n.insert(c);
    }

    for &digit_to_replace_char in &unique_digits_in_n {
        let mut prime_count = 0;
        for i in 0..=9 { // Replacement digit
            let replacement_digit_char = std::char::from_digit(i, 10).unwrap();

            let mut temp_s_chars = chars.clone();
            let mut replaced_at_first_pos = false;
            let mut replaced_any = false;

            for k in 0..len {
                if temp_s_chars[k] == digit_to_replace_char {
                    if k == 0 { // Check if the first digit is being replaced
                        replaced_at_first_pos = true;
                    }
                    temp_s_chars[k] = replacement_digit_char;
                    replaced_any = true;
                }
            }

            if !replaced_any {
                continue;
            }

            if replaced_at_first_pos && replacement_digit_char == '0' && len > 1 {
                continue;
            }

            let temp_s: String = temp_s_chars.into_iter().collect();
            let num = temp_s.parse::<u32>().unwrap();

            if primes_set.contains(&num) {
                prime_count += 1;
            }
        }
        if prime_count == 8 {
            return true;
        }
    }
    false
}

fn main() {
    let limit = 10_000_000;
    let primes_vec = sieve_of_eratosthenes(limit);
    let primes_set: HashSet<u32> = primes_vec.into_iter().collect();

    for prime in primes_set.iter().cloned() {
        // Skip single-digit primes as they cannot form families by replacing digits
        if prime < 10 {
            continue;
        }
        if digit_replace(prime, &primes_set) {
            println!("{}", prime);
            break; // Found the smallest
        }
    }
}
